### PR TITLE
Java: Add JDBC connection SSRF sinks

### PR DESCRIPTION
**Pull request by p0wn4j**
Add sinks for SSRF in case of when attacker can control JDBC connection string.
Sinks are:

1. JDBC DriverManager.getConnection
2. JDBC Driver.connect
3. HikariCP HikariConfig
4. SpringBoot DataSourceBuilder
5. Spring JDBC DriverManagerDataSource
6. JDBI

### CODE REVIEW

File: java/ql/lib/semmle/code/java/security/RequestForgery.qll [hunk 1]
```diff
@@ -33,13 +34,31 @@ private class DefaultRequestForgeryAdditionalTaintStep extends RequestForgeryAdd
   }
 }
 
+private class TypePropertiesRequestForgeryAdditionalTaintStep extends RequestForgeryAdditionalTaintStep {
+  override predicate propagatesTaint(DataFlow::Node pred, DataFlow::Node succ) {
+    exists(MethodAccess ma |
+      // Properties props = new Properties();  
+      // props.setProperty("jdbcUrl", tained);
+      // Propogated tainted value to the qualifier
```
Comment (smowton) [inline java/ql/lib/semmle/code/java/security/RequestForgery.qll:42]:
```suggestion
      // Properties props = new Properties();  
      // props.setProperty("jdbcUrl", tainted);
      // Propagate tainted value to the qualifier `props`
```

File: java/ql/lib/semmle/code/java/security/RequestForgery.qll [hunk 2]
```diff
@@ -33,13 +34,31 @@ private class DefaultRequestForgeryAdditionalTaintStep extends RequestForgeryAdd
...
+      ma.getMethod() instanceof PropertiesSetPropertyMethod and
+      ma.getArgument(0).(CompileTimeConstantExpr).getStringValue().matches("jdbcUrl") and
```
Comment (smowton) [inline java/ql/lib/semmle/code/java/security/RequestForgery.qll:44]:
```suggestion
      ma.getArgument(0).(CompileTimeConstantExpr).getStringValue() = "jdbcUrl" and
```

File: java/ql/lib/semmle/code/java/dataflow/ExternalFlow.qll [hunk 3]
```diff
@@ -265,6 +265,24 @@ private predicate sinkModelCsv(string row) {
       "java.net;URLClassLoader;false;URLClassLoader;(String,URL[],ClassLoader);;Argument[1];open-url",
       "java.net;URLClassLoader;false;URLClassLoader;(String,URL[],ClassLoader,URLStreamHandlerFactory);;Argument[1];open-url",
       "java.net;URLClassLoader;false;newInstance;;;Argument[0];open-url",
+      //JDBC URL
```
Comment (smowton) [inline java/ql/lib/semmle/code/java/dataflow/ExternalFlow.qll:268]: These should be moved to qll files relevant to their particular framework, and use an extension class, like

```
private class ExampleClassName extends SinkModelCsv {
  override predicate row(string row) {
    row =
      [
         ... row data here ...
      ]
  }
}
```

For example, in Jdbc.qll you'll see `private class SqlSinkCsv extends SinkModelCsv {`

Note that whenever a qll defines models this way you should add an import for it in `module Frameworks` on line 79 of this file.

File: java/ql/lib/semmle/code/java/dataflow/ExternalFlow.qll [hunk 4]
```diff
@@ -265,6 +265,24 @@ private predicate sinkModelCsv(string row) {
...
+      "java.sql;DriverManager;false;getConnection;(String);;Argument[0];jdbc-url",
+      "java.sql;DriverManager;false;getConnection;(String,Properties);;Argument[0];jdbc-url",
+      "java.sql;DriverManager;false;getConnection;(String,String,String);;Argument[0];jdbc-url",
+      "java.sql;Driver;false;connect;(String,Properties);;Argument[0];jdbc-url",
```
Comment (smowton) [inline java/ql/lib/semmle/code/java/dataflow/ExternalFlow.qll:272]: Move to Jdbc.qll

File: java/ql/lib/semmle/code/java/dataflow/ExternalFlow.qll [hunk 5]
```diff
@@ -265,6 +265,24 @@ private predicate sinkModelCsv(string row) {
...
+      "com.zaxxer.hikari;HikariConfig;false;HikariConfig;(Properties);;Argument[0];jdbc-url",
+      "com.zaxxer.hikari;HikariConfig;false;setJdbcUrl;(String);;Argument[0];jdbc-url",
```
Comment (smowton) [inline java/ql/lib/semmle/code/java/dataflow/ExternalFlow.qll:274]: Make a new qll file

File: java/ql/lib/semmle/code/java/dataflow/ExternalFlow.qll [hunk 6]
```diff
@@ -265,6 +265,24 @@ private predicate sinkModelCsv(string row) {
...
+      "org.springframework.boot.jdbc;DataSourceBuilder;false;url;(String);;Argument[0];jdbc-url",
+      "org.springframework.jdbc.datasource;AbstractDriverBasedDataSource;false;setUrl;(String);;Argument[0];jdbc-url",
+      "org.springframework.jdbc.datasource;DriverManagerDataSource;false;DriverManagerDataSource;(String);;Argument[0];jdbc-url",
+      "org.springframework.jdbc.datasource;DriverManagerDataSource;false;DriverManagerDataSource;(String,String,String);;Argument[0];jdbc-url",
+      "org.springframework.jdbc.datasource;DriverManagerDataSource;false;DriverManagerDataSource;(String,Properties);;Argument[0];jdbc-url",
```
Comment (smowton) [inline java/ql/lib/semmle/code/java/dataflow/ExternalFlow.qll:279]: Move to SpringJdbc.qll

File: java/ql/lib/semmle/code/java/dataflow/ExternalFlow.qll [hunk 7]
```diff
@@ -265,6 +265,24 @@ private predicate sinkModelCsv(string row) {
...
+      "org.jdbi.v3.core;Jdbi;false;create;(String);;Argument[0];jdbc-url",
+      "org.jdbi.v3.core;Jdbi;false;create;(String,Properties);;Argument[0];jdbc-url",
+      "org.jdbi.v3.core;Jdbi;false;create;(String,String,String);;Argument[0];jdbc-url",
+      "org.jdbi.v3.core;Jdbi;false;open;(String);;Argument[0];jdbc-url",
+      "org.jdbi.v3.core;Jdbi;false;open;(String,Properties);;Argument[0];jdbc-url",
+      "org.jdbi.v3.core;Jdbi;false;open;(String,String,String);;Argument[0];jdbc-url",
```
Comment (smowton) [inline java/ql/lib/semmle/code/java/dataflow/ExternalFlow.qll:285]: Make a new QLL file

**Comment by smowton** [comment]
Looks good -- I added a change note (take a look at its format for future reference), will merge on green.

**Comment by p0wn4j** [comment]
Thanks for the review :)

**Pull request merged**